runtime/nvqir/CircuitSimulator.h

```diff
@@ -626,20 +709,28 @@ class CircuitSimulatorBase : public CircuitSimulator {
       executionContext->simulationData = getStateData();
     }
 
-    executionContext = nullptr;
-
     // Deallocate the deferred qubits, but do so
     // without explicit qubit reset.
     for (auto &deferred : deferredDeallocation)
       tracker.returnIndex(deferred);
 
+    bool shouldSetToZero = isInBatchMode() && !isLastBatch();
+    executionContext = nullptr;
+
     // Reset the state if we've deallocated all qubits.
     if (tracker.numAvailable() == tracker.totalNumQubits()) {
-      cudaq::info("Deallocated all qubits, reseting state vector.");
-      // all qubits deallocated,
-      resetQubitState();
+      if (shouldSetToZero) {
+        cudaq::info("In batch mode currently, reset state to |0>");
+        // Do not deallocate the state, but reset it to |0>
+        setToZeroState();
+      } else {
+        cudaq::info("Deallocated all qubits, reseting state vector.");
+        // all qubits deallocated,
+        resetQubitState();
```

Comment:
Maybe a bit out-of-scope of this PR, but we should rename this function to `deallocateState` or something of the sort

Reply:
I like that suggestion